cli/flox-rust-sdk/src/providers/publish.rs

```diff
@@ -60,48 +97,57 @@ pub struct PublishProvider<Builder> {
     /// Directory under which we will clone and build the
     _base_temp_dir: PathBuf,
     /// Token of the user to authenticate with the catalog
-    _auth_token: FloxhubToken,
+    _auth_token: Option<FloxhubToken>,
     /// Building of manifest packages
     _builder: Builder,
 }
 
-#[derive(Debug, Error)]
-pub enum PublishError {
-    #[error("This type of environment is not supported for publishing")]
-    UnsupportedEnvironment,
-}
-
 /// (default) implementation of the `Publish` trait, i.e. the publish interface to publish.
 impl<Builder> Publish for PublishProvider<&Builder>
 where
     Builder: ManifestBuilder,
 {
     fn publish(
         &self,
+        flox: &Flox,
         _client: Client,
-        builder: impl ManifestBuilder,
         environment: PublishEnvironment,
-        _package: String,
-    ) -> Result<(), String> {
+        builder: impl ManifestBuilder,
+        package: &str,
+    ) -> Result<(), PublishError> {
+        // Get metadata from the environment, like locked URLs.
```

Comment:
question: why are managed environment unsupported?

changing _the signature of `check_environment_metadata` and the `gather_*` functions from `environment: &impl Environment,` to `environment: &impl Environment` compiles fine, meaning the metadata gathering does not depend on any inherent apis of the `PathEnvironment`

Reply:
suggestion:

I wonder if we need to depend on the _actual_ `Environment` implementation at all -- it doesnt look like it _yet_, but maybe worth discussing.

What I'm thinking about is changing `PublishEnvironment` to look more like this:


```rust
struct PublishEnvironment {
    parent_path: PathBuf, // required for building
    lockfile: Lockfile,
    rendered_environment: PathBuf // the store path of the built environment,
                                  // this will be a bit more elaborate after the buildenv changes
    git: Git                      // we dont need to read all the git metadata just yet,
                                  // but extract the git handle to start reading
}

impl PublishEnvironment {
    pub fn from_path_environment(flox: &Flox, env: PathEnvironment) -> Result<Self> {

    }
    pub fn from_managed_environment(flox: &Flox, env: ManagedEnvironment) -> Result<Self> {

    }

    // private(ish), as we dont want to create these for remote environments
    fn from_environment_unchecked(flox: &Flox, env: impl Environment) -> Result<Self> {
        let parent_path = env.parent_path();
        let git = /* find containing git repo or err */
        let lockfile = /* get lockfile with `Flox` or err */
        let rendered_environment = /* build environment or err */

        Ok(Self {
            parent_path,
            git,
            lockfile,
            rendered_environment,
         })
    }
}
```

As a side effect this also seems to cover most cases that currently depend on `Flox` (sans `build` which only reads runtime_dir off of it but we can change that i think)

Reply:
> question: why are managed environment unsupported?
> 
> changing _the signature of `check_environment_metadata` and the `gather_*` functions from `environment: &impl Environment,` to `environment: &impl Environment` compiles fine, meaning the metadata gathering does not depend on any inherent apis of the `PathEnvironment`

Just taking baby steps is all, just a 'not yet' thing as I work through implementation.  Reviewing your other comments and other changes may address this anyway.

Reply:
> I wonder if we need to depend on the _actual_ `Environment` implementation at all -- it doesnt look like it _yet_, but maybe worth discussing.

I don't have enough scope yet to know what things _will_ look like let alone where those things will come from.  We'll need to build and I was just getting into that code a bit in my second ticket.  If that's all possible with a model like you suggest above, I'm not against re-working it.  But I may ask that I address the other comments first, merge, and then have a follow up ticket to avoid long outstanding PRs.  Since this code isn't tied into anything at all yet, it shouldn't affect much else.

Reply:
That's totally fair, if that's only an intermediate step on the way.